plugin/src/HeadlessAPI.lua

```diff
@@ -0,0 +1,452 @@
+			-- No more than LIMIT requests per BUCKET seconds
+			return true
+		end
+
+		Rojo._rateLimit[source][api] += 1
+		task.delay(BUCKET, function()
+			Rojo._rateLimit[source][api] -= 1
+		end)
+
+		return false
+	end
+
+	Rojo._permissionsChangedEvent = Instance.new("BindableEvent")
+	Rojo._permissionsChanged = Rojo._permissionsChangedEvent.Event
+
+	function Rojo:_permissionCheck(key: string): boolean
+		if apiPermissionAllowlist[key] then
+			return true
+		end
+
+		local source = Rojo:_getCallerSource()
+		if Rojo._permissions[source] == nil then
+			Rojo._permissions[source] = {}
+		end
+
+		return not not Rojo._permissions[source][key]
+	end
+
+	function Rojo:_setPermissions(source, name, permissions)
+		-- Ensure permissions exist for this source
+		if Rojo._permissions[source] == nil then
+			Rojo._permissions[source] = {}
+		end
+
+		-- Set permissions
+		for api, granted in permissions do
+			Log.warn(
+				string.format(
+					"%s Rojo.%s for '%s'",
+					granted and "Granting permission to" or "Denying permission to",
+					api,
+					name
+				)
+			)
+			Rojo._permissions[source][api] = granted
+		end
+
+		-- Clear out source if no permissions are granted
+		local hasAnyPermissions = false
+		for _, granted in Rojo._permissions[source] do
+			if granted then
+				hasAnyPermissions = true
+				break
+			end
+		end
+		if not hasAnyPermissions then
+			Rojo._permissions[source] = nil
+		end
+
+		-- Update stored permissions
+		Settings:set("apiPermissions", Rojo._permissions)
+
+		-- Share changes
+		Rojo._permissionsChangedEvent:Fire(source, Rojo._permissions[source])
+	end
+
+	Rojo._apiDescriptions.RequestAccess = "Used to gain access to Rojo API members"
+	function Rojo:RequestAccess(plugin: Plugin, apis: { string }): { [string]: boolean }
+		assert(type(apis) == "table", "Rojo:RequestAccess expects an array of valid API names as the second argument")
+		assert(
+			typeof(plugin) == "Instance" and plugin:IsA("Plugin"),
+			"Rojo:RequestAccess expects a Plugin as the first argument"
+		)
+
+		local source, name = Rojo:_getCallerSource(), Rojo:_getCallerName()
+		Rojo._sourceToPlugin[source] = plugin
+
+		if Rojo:_checkRateLimit("RequestAccess") then
+			-- Because this opens a popup, we dont want to let users get spammed by it
+			return {}
+		end
+
+		if Rojo._activePermissionRequests[source] then
+			-- If a request is already active, exit
+			error(
+				"Rojo:RequestAccess cannot be called in multiple threads at once. Please call it once and wait for the response before calling it again.",
+				2
+			)
+		end
+		Rojo._activePermissionRequests[source] = true
+
+		if Rojo._permissions[source] == nil then
+			Rojo._permissions[source] = {}
+		end
+
+		-- Sanitize request
+		local sanitizedApis = {}
+		for _, api in apis do
+			if Rojo._apiDescriptions[api] ~= nil then
+				table.insert(sanitizedApis, api)
+			else
+				warn(string.format("Rojo.%s is not a valid API", tostring(api)))
+			end
+		end
```

Comment:
I looked at the code and I don't think it's an issue, but this should probably filter out duplicate valid APIs.